overlord/devicestate/handlers.go

```diff
@@ -790,3 +795,132 @@ func fetchKeys(st *state.State, keyID string) (errAcctKey error, err error) {
 		keyID = a.SignKeyID()
 	}
 }
+
+func snapState(st *state.State, name string) (*snapstate.SnapState, error) {
+	var snapst snapstate.SnapState
+	err := snapstate.Get(st, name, &snapst)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	return &snapst, nil
+}
+
+func makeRollbackDir(name string) (string, error) {
+	rollbackDir := filepath.Join(dirs.SnapRollbackDir, name)
+
+	if err := os.MkdirAll(rollbackDir, 0755); err != nil {
+		return "", err
+	}
+
+	return rollbackDir, nil
+}
+
+func currentGadgetInfo(snapst *snapstate.SnapState) (*gadget.Info, error) {
+	var gi *gadget.Info
+
+	currentInfo, err := snapst.CurrentInfo()
+	if err != nil && err != snapstate.ErrNoCurrent {
+		return nil, err
+	}
+	if currentInfo != nil {
+		gi, err = snap.ReadGadgetInfo(currentInfo, false)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return gi, nil
+}
+
+func pendingGadgetInfo(snapsup *snapstate.SnapSetup) (*gadget.Info, error) {
+	info, err := snap.ReadInfo(snapsup.InstanceName(), snapsup.SideInfo)
+	if err != nil {
+		return nil, err
+	}
+	update, err := snap.ReadGadgetInfo(info, false)
+	if err != nil {
+		return nil, err
+	}
+	return update, nil
+}
+
+func gadgetCurrentAndUpdate(st *state.State, snapsup *snapstate.SnapSetup) (current *gadget.Info, update *gadget.Info, err error) {
+	snapst, err := snapState(st, snapsup.InstanceName())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	currentInfo, err := currentGadgetInfo(snapst)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if currentInfo == nil {
+		// don't bother reading update if there is no current
+		return nil, nil, nil
+	}
+
+	newInfo, err := pendingGadgetInfo(snapsup)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return currentInfo, newInfo, nil
+}
+
+var (
+	gadgetUpdate = nopGadgetOp
+)
+
+func nopGadgetOp(current, update *gadget.Info, rollbackRootDir string) error {
+	return nil
+}
+
+func (m *DeviceManager) doUpdateGadget(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
+		return fmt.Errorf("cannot run update gadget task on a classic system")
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	snapsup, err := snapstate.TaskSnapSetup(t)
+	if err != nil {
+		return err
+	}
+
+	current, update, err := gadgetCurrentAndUpdate(t.State(), snapsup)
+	if err != nil {
+		return err
+	}
+	if current == nil {
+		// XXX: no updates during first boot & seeding
```

Comment:
That sounds right, ubuntu-image should have setup things properly at that point, not need for the XXX. UC20 story might be more complicated but we'll deal with it later.